### ISSUE TITLE
[torch.compile][Inductor] Silent wrong results when an automatically dynamic Python float used by clamp reuses a stale value

### ISSUE THREAD

**Issue by tongke6** [pytorch/pytorch#194976]
### Summary

`torch.compile(backend="inductor")` can silently reuse a stale Python float value after Dynamo automatically makes that value dynamic.

The issue affects both forward and backward computations. It reproduces with:

- PyTorch 2.8-based NVIDIA builds on CUDA
- The latest official PyTorch nightly on CPU Inductor
- Separate compiled closures created for each float value

The same reproducer is correct with:

- `torch._dynamo.config.patch(specialize_float=True)`
- `backend="aot_eager"`

The fact that the latest CPU Inductor reproduces the same numerical error indicates that this is not specific to CUDA, Triton, or a particular GPU kernel.

### Minimal reproducer

```python
import torch
import torch.nn.functional as F


def eager_swiglu(x, limit):
    gate, up = torch.chunk(x, 2, dim=-1)
    gate = F.silu(gate).clamp(max=limit)
    up = up.clamp(min=-limit, max=limit)
    return gate * up


@torch.compile(backend="inductor")
def compiled_swiglu(x, limit):
    gate, up = torch.chunk(x, 2, dim=-1)
    gate = F.silu(gate).clamp(max=limit)
    up = up.clamp(min=-limit, max=limit)
    return gate * up


print("torch:", torch.__version__)
print("git_version:", torch.version.git_version)
print("specialize_float:", torch._dynamo.config.specialize_float)

device = "cuda" if torch.cuda.is_available() else "cpu"

cases = [
    ((1, 2), False, 1.0),
    ((1, 4), True, 0.5),
    ((1, 4), True, 1.0),
]

torch._dynamo.reset()

with torch._dynamo.config.patch(specialize_float=False):
    for shape, requires_grad, limit in cases:
        x = torch.full(
            shape,
            0.75,
            device=device,
            dtype=torch.float32,
            requires_grad=requires_grad,
        )
        x_ref = x.detach().clone().requires_grad_(requires_grad)

        actual = compiled_swiglu(x, limit)
        expected = eager_swiglu(x_ref, limit)

        forward_diff = (actual - expected).abs().max().item()
        grad_diff = None

        if requires_grad:
            actual.sum().backward()
            expected.sum().backward()
            grad_diff = (x.grad - x_ref.grad).abs().max().item()

        print(
            f"shape={shape}, requires_grad={requires_grad}, limit={limit}, "
            f"forward_diff={forward_diff}, grad_diff={grad_diff}"
        )
```

The invocation order is important:

1. `shape=(1, 2)`, `requires_grad=False`, `limit=1.0`
2. `shape=(1, 4)`, `requires_grad=True`, `limit=0.5`
3. `shape=(1, 4)`, `requires_grad=True`, `limit=1.0`

On the third invocation, Inductor may execute code containing the stale value `0.5`, even though the runtime argument is `1.0`.

### Expected behavior

All calls should match eager execution:

```text
forward_diff=0.0
grad_diff=0.0
```

When `specialize_float=False`, the automatically dynamic Python float should either:

1. remain genuinely dynamic throughout the compiled graph, or
2. be safely specialized with a value guard and recompiled when its value changes.

It must not be silently embedded as an unguarded stale constant.

### Actual behavior on the latest official nightly

This was reproduced using the official CPU nightly wheel:

```text
torch: 2.15.0.dev20260826+cpu
git_version: 68a5d90dd35d45ca457fda21dbae885987fb1e05
python: 3.12.3
device: cpu
backend: inductor
specialize_float: False

shape=(1, 2), requires_grad=False, limit=1.0,
forward_diff=0.0, grad_diff=None

shape=(1, 4), requires_grad=True, limit=0.5,
forward_diff=0.0, grad_diff=0.0

shape=(1, 4), requires_grad=True, limit=1.0,
forward_diff=0.13203802704811096,
grad_diff=0.2106499969959259
```

This is a silent correctness issue: no exception or warning is raised.

The CPU forward error is exactly the same as the previously observed CUDA forward error:

```text
0.13203802704811096
```

#### `specialize_float=True`: correct

Changing the configuration to:

```python
with torch._dynamo.config.patch(specialize_float=True):
```

produces:

```text
shape=(1, 2), requires_grad=False, limit=1.0,
forward_diff=0.0, grad_diff=None

shape=(1, 4), requires_grad=True, limit=0.5,
forward_diff=0.0, grad_diff=0.0

shape=(1, 4), requires_grad=True, limit=1.0,
forward_diff=0.0, grad_diff=0.0
```

#### `aot_eager`: correct

Changing the backend to:

```python
@torch.compile(backend="aot_eager")
```

while keeping `specialize_float=False` also produces zero forward and gradient errors.

This narrows the issue to the Inductor/AOTAutograd scalar-tensorification path rather than Dynamo tracing alone.

### Closure variant

Creating a new compiled callable for every `limit` value does not avoid the bug:

```python
def make_compiled_swiglu(limit):
    @torch.compile(backend="inductor")
    def compiled_swiglu(x):
        gate, up = torch.chunk(x, 2, dim=-1)
        gate = F.silu(gate).clamp(max=limit)
        up = up.clamp(min=-limit, max=limit)
        return gate * up

    return compiled_swiglu
```

Running the same sequence with a newly created closure for every case still gives, on the latest nightly:

```text
shape=(1, 4), requires_grad=True, limit=1.0,
forward_diff=0.13203802704811096,
grad_diff=0.2106499969959259
```

With `specialize_float=True`, the closure variant is correct.

The nested functions share the same Python code object, so the Dynamo automatic-dynamic state appears to be shared even though the callable objects and closure cells are distinct.

### Suspected root cause

The suspected issue is in:

```text
torch/fx/passes/_tensorify_python_scalars.py
```

In the current implementation:

- only a limited set of operators is registered in `SUPPORTED_OPS`;
- `aten.clamp.default`, `aten.clamp_min.default`, and `aten.clamp_max.default` are not tensorified;
- scalar-use detection and transformation primarily inspect positional `node.args`;
- `clamp(max=limit)` and `clamp(min=-limit, max=limit)` carry their bounds in `node.kwargs`;
- the same backed `SymFloat` can be marked as tensorified because some of its other uses are tensorifiable;
- the final specialization sweep can then skip that symbol even though a non-tensorified clamp use still requires specialization.

This may leave the clamp bound embedded as an unguarded trace-time constant, allowing a graph compiled with `limit=0.5` to be reused when `limit=1.0`.

The bug is related to how `clamp` exposes the missing mixed-use handling, but does not appear to be a numerical bug in the clamp implementation itself.

A correct fix should preserve the original intent of `specialize_float=False`:

- tensorify all supported uses of the backed `SymFloat`, including keyword arguments; and
- if any use cannot be tensorified, specialize and guard that symbol rather than treating it as fully tensorified.

Adding clamp tensorification would keep this common case genuinely dynamic, while the mixed-use specialization rule would provide a correctness fallback for other unsupported operators.

### Device/backend observations

The issue has now been reproduced with both:

- CUDA Inductor in the original production environment
- CPU Inductor in the latest official nightly

Therefore, the root cause appears to be in the device-independent Dynamo/AOTAutograd/Inductor graph transformation logic, before CPU C++ or CUDA/Triton code generation.

Changing the Triton version did not fix the original CUDA reproduction.

### Debugging

Generated-code and recompilation logs can be collected with:

```bash
TORCH_LOGS="recompiles,output_code" python repro.py
```

The relevant symptom is that the third invocation passes `limit=1.0`, while generated or reused Inductor code still contains the previous value `0.5`.

### Error logs

There is no exception. This is a silent correctness failure.

Latest official nightly, CPU Inductor:

torch: 2.15.0.dev20260826+cpu
git_version: 68a5d90dd35d45ca457fda21dbae885987fb1e05
backend: inductor
specialize_float: False

shape=(1, 2), requires_grad=False, limit=1.0, forward_diff=0.0, grad_diff=None
shape=(1, 4), requires_grad=True, limit=0.5, forward_diff=0.0, grad_diff=0.0
shape=(1, 4), requires_grad=True, limit=1.0, forward_diff=0.13203802704811096, grad_diff=0.2106499969959259

### Versions

Latest nightly reproduction:

PyTorch version: 2.15.0.dev20260826+cpu
PyTorch git version: 68a5d90dd35d45ca457fda21dbae885987fb1e05
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: 11.4.0
Clang version: 16.0.4
CMake version: 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3
Python platform: Linux-6.8.0-58-generic-x86_64-with-glibc2.39
Is CUDA available: False

CPU: AMD Ryzen Threadripper PRO 5975WX 32-Cores

Relevant libraries:
[pip3] torch==2.15.0.dev20260826+cpu
[conda] Could not collect

cc @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo
